improver/utilities/cube_extraction.py

```diff
@@ -250,3 +252,131 @@ def extract_subcube(cube, constraints, units=None, use_original_units=True):
+        cutout = cube.extract(lat_constraint)
+        if cutout is None:
+            return cutout
+
+        cutout = cutout.intersection(longitude=(xmin, xmax), ignore_bounds=True)
+
+        # intersection creates a new coordinate with default datatype - we
+        # therefore need to re-cast to meet the IMPROVER standard
+        cutout.coord("longitude").points = cutout.coord("longitude").points.astype(
+            FLOAT_DTYPE
+        )
+        if cutout.coord("longitude").bounds is not None:
+            cutout.coord("longitude").bounds = cutout.coord("longitude").bounds.astype(
+                FLOAT_DTYPE
+            )
+
+    else:
+        x_constraint = Constraint(
+            projection_x_coordinate=lambda x: xmin <= x.point <= xmax
+        )
+        y_constraint = Constraint(
+            projection_y_coordinate=lambda x: ymin <= x.point <= ymax
+        )
+        cutout = cube.extract(x_constraint & y_constraint)
+
+    return cutout
+
+
+def subset_data(cube, grid_spec=None, site_list=None):
+    """Extract a spatial cutout or subset of sites from data
+    to generate suite reference outputs.
+
+    Args:
+        cube (iris.cube.Cube):
+            Input dataset
+        grid_spec (Optional[Dict[str, Dict[str, int]]]):
+            Dictionary containing bounding grid points and an integer "thinning
+            factor" for each of UK and global grid, to create cutouts.  Eg a
+            "thinning factor" of 10 would mean every 10th point being taken for
+            the cutout.  The expected dictionary has keys that are spatial coordinate
+            names, with values that are dictionaries with "min", "max" and "thin" keys.
+        site_list (Optional[list]):
+            List of WMO site IDs to extract.  These IDs must match the type and format
+            of the "wmo_id" coordinate on the input spot cube.
+
+    Returns:
+        iris.cube.Cube:
+            Subset of input cube as specified by input constraints
+
+    Raises:
+        ValueError:
+            If site_list is not provided for a spot data cube
+        ValueError:
+            If the spot data cube does not contain any of the required sites
+        ValueError:
+            If grid_spec is not provided for a gridded cube
+        ValueError:
+            If grid_spec does not contain entries for the spatial coordinates on
+            the input gridded data
+        ValueError:
+            If the grid_spec provided does not overlap with the cube domain
+    """
+    if cube.coords("spot_index"):
+        if site_list is None:
+            raise ValueError("site_list required to extract from spot data")
+
+        constraint = Constraint(coord_values={"wmo_id": lambda x: x in site_list})
+        result = cube.extract(constraint)
+        if result is None:
+            raise ValueError(
+                f"Cube does not contain any of the required sites: {site_list}"
+            )
```

Comment:
I'm anticipating automated usage where the grid spec is going to be defined for a range of input files, which might not all contain the same sites or data.  So I didn't want to raise anything here as long as there was some overlap.

Can you envisage a case in which a warning would be helpful?  (Bearing in mind that they're not helpful operationally - so this would be manual code use of the CLI for some reason.)

Reply:
Thinking again, as this is all targeted at the automated tests as you point out, I think you're correct that a warning is not needed.